Dockerfile

```diff
@@ -45,6 +47,20 @@ ENV CARGO_HOME=/cargo \
 # (keep version in sync with src/internet_identity/build.sh)
 RUN cargo install ic-cdk-optimizer --version 0.3.1
 
+# warm up the cache. Since cargo doesn't have a flag to only build the
+# dependecies, we copy all the files (Cargo.toml, Cargo.lock, vendored deps)
+# _except_ for the actual internet identity code, which we replace with a dummy.
+COPY Cargo.lock .
+COPY Cargo.toml .
+COPY src/cubehash src/cubehash
+COPY src/internet_identity/Cargo.toml src/internet_identity/Cargo.toml
+# NOTE: we make the timestamp of src/lib.rs very old so that cargo knows to
+# rebuild the real file when we copy it
+RUN mkdir -p src/internet_identity/src && touch -t 197001010000 src/internet_identity/src/lib.rs
+RUN cargo build --target wasm32-unknown-unknown --release -j1
+
+# copy the actual code
+RUN rm -rf src
```

Comment:
Do we want to combine those into a single RUN?

Reply:
Unless we have too many layers, I'd rather optimize for legibility, but up to you!